daemons/fenced/fenced_scheduler.c

```diff
@@ -97,119 +97,107 @@ local_node_allowed_for(const pcmk_resource_t *rsc)
     return NULL;
 }
 
-#define rsc_name(x) x->clone_name?x->clone_name:x->id
-
 /*!
  * \internal
- * \brief If a resource or any of its children are STONITH devices, update their
- *        definitions given a cluster working set.
+ * \brief If a given resource or any of its children are fencing devices,
+ *        register the devices
  *
  * \param[in,out] data       Resource to check
- * \param[in,out] user_data  Cluster working set with device information
+ * \param[in,out] user_data  Ignored
  */
 static void
-cib_device_update(pcmk_resource_t *rsc, pcmk_scheduler_t *data_set)
+register_if_fencing_device(gpointer data, gpointer user_data)
 {
+    pcmk_resource_t *rsc = data;
+
+    xmlNode *xml = NULL;
+    GHashTableIter hash_iter;
     pcmk_node_t *node = NULL;
+    const char *name = NULL;
     const char *value = NULL;
     const char *rclass = NULL;
-    pcmk_node_t *parent = NULL;
-
-    /* If this is a complex resource, check children rather than this resource itself. */
-    if(rsc->children) {
-        GList *gIter = NULL;
-        for (gIter = rsc->children; gIter != NULL; gIter = gIter->next) {
-            cib_device_update(gIter->data, data_set);
-            if(pe_rsc_is_clone(rsc)) {
-                crm_trace("Only processing one copy of the clone %s", rsc->id);
-                break;
+    const char *agent = NULL;
+    const char *rsc_provides = NULL;
+    stonith_key_value_t *params = NULL;
+
+    // If this is a collective resource, check children instead
+    if (rsc->children != NULL) {
+        for (GList *iter = rsc->children; iter != NULL; iter = iter->next) {
+            register_if_fencing_device(iter->data, NULL);
+            if (pe_rsc_is_clone(rsc)) {
+                return; // Only one instance needs to be checked for clones
             }
         }
         return;
     }
 
-    /* We only care about STONITH resources. */
     rclass = crm_element_value(rsc->xml, XML_AGENT_ATTR_CLASS);
     if (!pcmk__str_eq(rclass, PCMK_RESOURCE_CLASS_STONITH, pcmk__str_casei)) {
```

Comment:
Do we need to compare case insensitive here?

Reply:
We probably *shouldn't*, but I don't want to break any existing configurations. For provider and agent it's easier to insist on case sensitivity, but class names are an invention of Pacemaker. There were probably variations in the early days that might still survive somewhere. We could set a stricter policy at a backward compatibility break.